File: x/go/border/border.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package border

import (
	"encoding/json"

	"github.com/synnaxlabs/x/spatial"
)

// UnmarshalJSON supports four formats:
//   - number: 4
//   - pair:   {"x": 4, "y": 8}
//   - per-corner numbers: {"top_left": 4, ...}
//   - per-corner pairs:   {"top_left": {"x": 4, "y": 8}, ...}
//
// Corner keys are accepted in both snake_case and camelCase, since legacy
// schematic configs stored them verbatim from the Console. Shorthand forms
// expand to the canonical per-corner shape.
func (r *Radius) UnmarshalJSON(data []byte) error {
	if string(data) == "null" {
		*r = Radius{}
		return nil
	}
	var n float64
	if err := json.Unmarshal(data, &n); err == nil {
		v := spatial.XY{X: n, Y: n}
		*r = Radius{TopLeft: v, TopRight: v, BottomLeft: v, BottomRight: v}
		return nil
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(data, &obj); err != nil {
		return err
	}
	if _, hasX := obj["x"]; hasX {
		var v spatial.XY
		if err := json.Unmarshal(data, &v); err != nil {
			return err
		}
		*r = Radius{TopLeft: v, TopRight: v, BottomLeft: v, BottomRight: v}
		return nil
	}
	corners := []struct {
		snake, camel string
		dst          *spatial.XY
	}{
		{"top_left", "topLeft", &r.TopLeft},
		{"top_right", "topRight", &r.TopRight},
		{"bottom_left", "bottomLeft", &r.BottomLeft},
		{"bottom_right", "bottomRight", &r.BottomRight},
	}
	for _, c := range corners {
		raw, ok := obj[c.snake]
		if !ok {
			raw, ok = obj[c.camel]
		}
		if !ok {
			continue
		}
		var num float64
		if err := json.Unmarshal(raw, &num); err == nil {
			*c.dst = spatial.XY{X: num, Y: num}
			continue
		}
		if err := json.Unmarshal(raw, c.dst); err != nil {
			return err
		}
	}
	return nil
}

File: x/ts/src/border/border.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { z } from "zod";

import { xy } from "@/spatial/xy";

/** Zod schema for {@link Radius}. */
export const radiusZ = z.object({
  topLeft: xy.xyZ,
  topRight: xy.xyZ,
  bottomLeft: xy.xyZ,
  bottomRight: xy.xyZ,
});

/**
 * Per-corner elliptical radii, each corner carrying an independent x and y component.
 * The caller sets the units, whether pixels or percentages of the element.
 */
export interface Radius extends z.infer<typeof radiusZ> {}

const numberCornersZ = z.object({
  topLeft: z.number(),
  topRight: z.number(),
  bottomLeft: z.number(),
  bottomRight: z.number(),
});

/** Zod schema for {@link CrudeRadius}. */
export const crudeRadiusZ = z.union([z.number(), xy.xyZ, numberCornersZ, radiusZ]);

/** A {@link Radius} in any of its shorthand forms. */
export type CrudeRadius = z.infer<typeof crudeRadiusZ>;

/**
 * @constructs Radius
 * @param radius - A scalar or a single {x, y} applied to every corner, or per-corner
 * numbers or {x, y} pairs.
 */
export const constructRadius = (radius: CrudeRadius): Radius => {
  if (typeof radius === "number" || "x" in radius) {
    const corner = xy.construct(radius);
    return {
      topLeft: corner,
      topRight: corner,
      bottomLeft: corner,
      bottomRight: corner,
    };
  }
  return {
    topLeft: xy.construct(radius.topLeft),
    topRight: xy.construct(radius.topRight),
    bottomLeft: xy.construct(radius.bottomLeft),
    bottomRight: xy.construct(radius.bottomRight),
  };
};
